### DISCUSSION TITLE
Current macOS support

### DISCUSSION THREAD

**Discussion by wq9578** [fail2ban/fail2ban#3788] [Q&A]
Does fail2ban support current macOS?
There is [a Homebrew formula](https://formulae.brew.sh/formula/fail2ban#default), but it seems that currently IP addresses of failed login attempts are [not logged in recent Apple systems](https://apple.stackexchange.com/a/322363).
I'm getting these `com.apple.xpc.launchd[1] (com.openssh.sshd._hex_[_int_]): Service exited with abnormal code: 255` messages (mentioned also at the top of the thread of the last link above).

**Comment by sebres** [comment]
> Does fail2ban support current macOS?

No idea... @fail2ban/maintainers can someone give a clue here, please?

> it seems that currently IP addresses of failed login attempts are not logged in recent Apple systems

Did you try to set VERBOSE level as described in the filter (dunno whether it is configurable on macOS):
https://github.com/fail2ban/fail2ban/blob/8170e9fe75fd2c2c4c51a1d9972b683401cddccb/config/filter.d/sshd.conf#L8

> I'm getting these `com.apple.xpc.launchd[1] (com.openssh.sshd._hex_[_int_]): Service exited with abnormal code: 255` messages

I found another answer on SE - https://apple.stackexchange.com/questions/171465/sshd-service-exited-with-abnormal-code-255

Anyway, it'll be impossible to match that messages (because there is no IP and no session-ID, one could bind it to some other message with IP using multi-line processing).

**Comment by wq9578** [comment]
I was able to find out more details. Details below.

**That still leaves open the question: Does fail2ban support current macOS?**

I have installed [OpenSSH via Homebrew](https://formulae.brew.sh/formula/openssh#default), however it seems that the LibreSSL `sshd` shipped with macOS is started when logging into the server, since that version (log: `Local version string SSH-2.0-OpenSSH_8.1 `) is used. My currently installed versions are:
```
$ which sshd; sshd -V               
/opt/homebrew/sbin/sshd
OpenSSH_9.2, OpenSSL 1.1.1s  1 Nov 2022
$ /usr/bin/ssh -V; /usr/sbin/sshd -V
OpenSSH_8.1p1, LibreSSL 2.7.3
/usr/sbin/sshd: illegal option -- V
OpenSSH_8.1p1, LibreSSL 2.7.3
usage: sshd [-46DdeiqTt] [-C connection_spec] [-c host_cert_file]
            [-E log_file] [-f config_file] [-g login_grace_time]
            [-h host_key_file] [-o option] [-p port] [-u len]
```

_Details_

Showing details about a failed connection, with `LogLevel DEBUG3` in `/etc/ssh/sshd_config`, live output via the command `log stream --level debug --style syslog`:
```
2024-07-08 21:40:14.[...]  localhost sshd[...]: debug3: BSM audit: connection from 9.9.9.9 port 38628
[...]
2024-07-08 21:40:14.[...]  localhost sshd[...]: Connection from 9.9.9.9 port 38628 on 1.1.1.1 port 22
[...]
2024-07-08 21:40:14.[...]  localhost sshd[...]: debug1: Local version string SSH-2.0-OpenSSH_8.1
2024-07-08 21:40:14.[...]  localhost sshd[...]: debug1: Remote protocol version 2.0, remote software version Go
[...]
2024-07-08 21:40:14.[...]  localhost sshd[...]: Unable to negotiate with 9.9.9.9 port 38628: [...] [preauth]
```


Or `log show --process sshd --info --last 10m`:
```
2024-07-08 21:47:56.[...] [...] Info        [...]                  [...]  [...]    sshd: Connection from 9.9.9.9 port 52514 on 1.1.1.1 port 22
2024-07-08 21:47:56.[...] [...] Info        [...]                  [...]  [...]    sshd: Unable to negotiate with 9.9.9.9 port 52514: [...] [preauth]
```
and
```
2024-07-08 22:48:35.[...] [...] Info        [...]                  [...]  [...]    sshd: Accepted publickey for root from 2.2.2.2 port 55715 ssh2: ECDSA SHA256:[...]
```

Without `LogLevel DEBUG3` (just the default `#LogLevel INFO`) still the relevant messages ("Unable to negotiate", "Accepted publickey for") appear when running `log stream` or `log show`.

Showing details about a failed login (successful connection):

`praudit -l /var/audit/current | grep ",failure"`
```
[...],OpenSSH login,[...],Mon Jul  8 22:07:21 2024, + 840 msec,[...],9.9.9.9,[...],too many tries for user root,return,failure : Operation not permitted,[...],com.apple.sshd,complete,,complete,[...],
```

Links:
* [Audit in a OS X System](https://www.scip.ch/en/?labs.20150108)
* [Quick and Dirty OpenBSM Auditing In macOS](https://krypted.com/mac-security/quick-dirty-openbsm-auditing-macos/)
